precompiles/staking/tx.go

```diff
@@ -365,3 +367,38 @@ func (p Precompile) CancelUnbondingDelegation(
 
 	return method.Outputs.Pack(true)
 }
+
+// Restake performs a restake of the provided delegation for the combination delegator <-> validator
+func (p Precompile) Restake(
+	ctx sdk.Context,
+	origin common.Address,
+	contract *vm.Contract,
+	stateDB vm.StateDB,
+	method *abi.Method,
+	args []interface{},
+) ([]byte, error) {
+	delegatorAddr, validatorAddr, err := checkRestakeArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	validator, found := p.stakingKeeper.GetValidator(ctx, validatorAddr.Bytes())
+	if !found {
+		return nil, fmt.Errorf("validator %s not found", validatorAddr.String())
+	}
+
+	delegation, found := p.stakingKeeper.GetDelegation(ctx, delegatorAddr.Bytes(), validatorAddr.Bytes())
+	if !found {
+		return nil, fmt.Errorf("delegation %s not found", delegatorAddr.String())
+	}
+
+	pendingRewards, err := p.distKeeper.WithdrawDelegationRewardsWithoutSending(ctx, validator, delegation)
+	if err != nil {
+		return nil, err
+	}
+
+	delegation.Shares = delegation.Shares.Add(pendingRewards[0].Amount.ToLegacyDec())
+	p.stakingKeeper.SetDelegation(ctx, delegation)
```

Comment:
@Vvaradinov we will need to emit the `Delegate` on the EVM and the Cosmos SDK too

Reply:
Calling `Delegate` through the msgServer throws a `no delegation distribution info` error because delegating calls into a hook that withdraws the rewards again hitting that error. The alternative is to either:
- Include the Cosmos Event emitting logic in the precompile call
- Remove the `!k.HasDelegatorStartingInfo(ctx, del.GetValidatorAddr(), del.GetDelegatorAddr())` check from the new `WithdrawDelegationRewards` method